File: lecture-practice-code/java-to-kotlin-library-app/src/main/java/com/group/javatokotlinlibraryapp/service/user/UserService.java
package com.group.javatokotlinlibraryapp.service.user;

import com.group.javatokotlinlibraryapp.domain.user.User;
import com.group.javatokotlinlibraryapp.domain.user.UserRepository;
import com.group.javatokotlinlibraryapp.dto.user.request.UserCreateRequest;
import com.group.javatokotlinlibraryapp.dto.user.request.UserUpdateRequest;
import com.group.javatokotlinlibraryapp.dto.user.response.UserResponse;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class UserService {

    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Transactional
    public void saveUser(UserCreateRequest request) {
        User newUser = new User(request.getName(), request.getAge());
        userRepository.save(newUser);
    }

    @Transactional(readOnly = true)
    public List<UserResponse> getUsers() {
        return userRepository.findAll().stream()
            .map(UserResponse::new)
            .collect(Collectors.toList());
    }

    @Transactional
    public void updateUserName(UserUpdateRequest request) {
        User user = userRepository.findById(request.getId()).orElseThrow(IllegalArgumentException::new);
        user.updateName(request.getName());
    }

    @Transactional
    public void deleteUser(String name) {
        User user = userRepository.findByName(name).orElseThrow(IllegalArgumentException::new);
        userRepository.delete(user);
    }

}

File: lecture-practice-code/java-to-kotlin-library-app/src/main/kotlin/com/group/javatokotlinlibraryapp/service/user/UserService.kt
package com.group.javatokotlinlibraryapp.service.user

import com.group.javatokotlinlibraryapp.domain.user.User
import com.group.javatokotlinlibraryapp.domain.user.UserRepository
import com.group.javatokotlinlibraryapp.dto.user.request.UserCreateRequest
import com.group.javatokotlinlibraryapp.dto.user.request.UserUpdateRequest
import com.group.javatokotlinlibraryapp.dto.user.response.UserResponse
import com.group.javatokotlinlibraryapp.util.fail
import com.group.javatokotlinlibraryapp.util.findByIdOrThrow
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserService(
    private val userRepository: UserRepository,
) {

    @Transactional
    fun saveUser(request: UserCreateRequest) {
        val newUser = User(request.name, request.age)
        userRepository.save(newUser)
    }

    @Transactional(readOnly = true)
    fun getUsers(): List<UserResponse> {
        return userRepository.findAll()
            .map { user -> UserResponse(user) }
    }

    @Transactional
    fun updateUserName(request: UserUpdateRequest) {
        val user = userRepository.findByIdOrThrow(request.id)
        user.updateName(request.name)
    }

    @Transactional
    fun deleteUser(name: String) {
        val user = userRepository.findByName(name) ?: fail()
        userRepository.delete(user)
    }
}